46.md

```diff
@@ -2,161 +2,83 @@ NIP-46
 ======
 
 Nostr Connect
-------------------------
+-------------
 
 `draft` `optional`
 
-## Rationale
+This NIP describes a method for 2-way communication between a **remote signer** and a normal Nostr client. The remote signer could be, for example, a hardware device dedicated to signing Nostr events, while the client is a normal Nostr client.
 
-Private keys should be exposed to as few systems - apps, operating systems, devices - as possible as each system adds to the attack surface.
+## Signer Discovery
 
-Entering private keys can also be annoying and requires exposing them to even more systems such as the operating system's clipboard that might be monitored by malicious apps.
+The client always starts by generating a random key which is used to communicate with the signer, then it one of the methods below is used to allow the client to know what is the signer public key for the session and which relays to use.
 
+### Started by the signer
 
-## Terms
-
-* **App**: Nostr app on any platform that *requires* to act on behalf of a nostr account.
-* **Signer**: Nostr app that holds the private key of a nostr account and *can sign* on its behalf.
+The remote signer generates a connection token in the form
 
+```
+<npub1...>#<secret>?relay=wss://...&relay=wss://...
+```
 
-## `TL;DR`
+The user copies that token and pastes it in the client UI somehow. Then the client can send events of kind `24133` to the specified relays and wait for responses from the remote signer.
 
+### Started by the client
 
-**App** and **Signer** sends ephemeral encrypted messages to each other using kind `24133`, using a relay of choice. 
+The client generates a QR code in the following form (URL-encoded):
 
-App prompts the Signer to do things such as fetching the public key or signing events.
+```
+nostrconnect://<client-key-hex>?relay=wss://...&metadata={"name":"...", "url": "...", "description": "..."}
+```
 
-The `content` field must be an encrypted JSONRPC-ish **request** or **response**.
+The signer scans the QR code and sends a `connect` message to the client in the specified relays.
```

Comment:
```suggestion
This link is input into the signer using one of the methods below:
- signer scanning a QR code
- deeplinking to the signer app
- pasting it into the signer

Once the link has been entered into the signer, the signer should send an event with payload of { id: "connect", result: `true` } using the indicated relays as acknowledgement that the signer is ready to receive requests from the client.
```
we can change the payload to whatever people think is best. Since the spec says that the payloads for events sent by the signer should look like {id: <request-id>, result: <any>, error: <reason-string>} it was confusing to say that the signer should send a 'connect' message. I figured its easier to just send a dummy initial connection payload. alternatively, we can change the spec to allow signer to send requests as well but I don't like that at all.

Reply:
Signer should be able to send the relay list in this 'connect' message for the app->signer flow. App can't know which relays work best with this signer, app chooses some random relay and puts into nostrconnect: url, signer sends 'connect' back with a list of proper relays as payload. App should then switch and use those relays instead of the random one (or at least add those relays to the pool).

General purpose relays don't work well with nip46 - you quickly hit rate limits for EVENT messages. Plus you can build specialized optimized relays for ephemeral-only load. Plus signer might actually be bundled with a relay to avoid a second network hop. Given all this, it's the signer that should decide which relays work best with it.

Reply:
I actually completely agree with this (about signer sending back relay with connect response). was going to suggest something similar but thought it would add more complexity and changes so I didn't bother. thank you for throwing it out there!

Reply:
Also, Nostr Wallet Auth proposal https://github.com/nostr-protocol/nips/pull/851/ is solving a similar flow of app->wallet, they too publish a 'connect'-like event after connection is confirmed in the wallet, but they use 30k replaceable event. Their justification is:
> We use a parameterized replaceable event to allow for the wallet and app to
renegotiate the connection if needed and to allow for the wallet being able to look up the connection details later.

Since this 'connect' confirmation event is a new idea in this rewrite maybe we could use a 30k event here too?

Reply:
I think they use 10k replaceable event, and it doesnt seem to contain information specific to the user (just holds what methods the wallet service supports). 

i'm not sure how comfortable I would be for this nostrConnect connect event to be queryable. 

it is interesting how different these two nips are despite some of their similarities. for example nip47 uses two different kinds for request and response, here we use one.

Reply:
They changed it to 30k, and it's encrypted so it only leaks metadata that a user was connecting to 'some app', so maybe it's not that bad.

Reply:
> it doesnt seem to contain information specific to the user

Their connect event does contain a relay, which is super useful.